[B+C] Add TeleportCause to EntityTeleportEvent. Adds BUKKIT-4745

Ribesg opened the pull request:
Based on PR by @Obliv
### The Issue:

EntityTeleportEvent does not have any information about teleport cause, while PlayerTeleportEvent does.
### Justification for this PR:

This standardizes the capabilities of teleport events between players and entities.
### PR Breakdown:
##### Bukkit

Added a teleport cause parameter to EntityTeleportEvent. Updated relevant calls to constructors in child classes.
##### CraftBukkit

Updated calls to EntityTeleportEvent and sub-events
### Testing Results and Materials:

Test Plugin: [Code on Github](https://github.com/Ribesg/TestPlugin/tree/BUKKIT-4745) - [Latest build on Jenkins](http://ci.ribesg.fr/view/Bukkit-PRs/job/PR-BUKKIT-4745/lastSuccessfulBuild/artifact/out/TestPlugin.jar)
We're not allowed to link CB builds here, you'll have to clone & build it youself:

``` bash
git clone https://github.com/Ribesg/Bukkit.git -b BUKKIT-4745 && mvn install
git clone https://github.com/Ribesg/CraftBukkit.git -b BUKKIT-4745 && mvn package
```
### Relevant PR(s):

CB-1311 - https://github.com/Bukkit/CraftBukkit/pull/1311 - Associated CraftBukkit PR
B-931 - https://github.com/Bukkit/Bukkit/pull/931 - Original PR
### JIRA Ticket:

BUKKIT-4745 - https://bukkit.atlassian.net/browse/BUKKIT-4745

Wolvereness commented:
File permissions are borked.

Ribesg commented:
Yeah I somehow fixed it at some point but I reverted that commit... Trying to fix it now (Windows, eh)

Ribesg commented:
@Wolvereness Fixed.

gmcferrin commented:
@Ribesg - Why not use a unified TeleportCause between both this and PlayerTeleportEvent? Is it really worth having two separate enums?

Ribesg commented:
@gmcferrin This would ever require to use the TeleportCause **in** PlayerTeleportEvent, which doesn't make much sense for me, or get TeleportCause out of PlayerTeleportEvent, which would be a breaking change.

Additionally, from a API user point of view, having a TeleportCause that contains only "real" possibilities is a good thing. For example, having to handle TeleportCause.ENDERMAN in a PlayerTeleportEvent (even as a default statement in a switch) would not make sense for me. Here the user only sees what's strictly linked to the event. That's my opinion on this.

I agree that some (most?) values match between the 2 enums. Anyway, I don't think that I'll be the one that will decide on this, but I'll be happy to change it the way you want... if you're ready to explain why it would be better :-)

gmcferrin commented:
Nope, your logic makes sense to me. Just wanted to hear your justification.

Ribesg committed:
Add TeleportCause to EntityTeleportEvent. Adds BUKKIT-4745

Ribesg commented:
I had to stash the commits to fix my repository. Sorry if it causes any problem.

Also updated to latest commit as of 2014-05-10.

nightpool commented:
@Ribesg Entity.teleport still seems to use a `PlayerTeleportEvent.TeleportCause`. This seems to tell me that we should have a unified TeleportCause api, so that Entity.teleport can have the same signature between players and entities.

Also, currently Entity.teleport doesn't throw an event (even though we have an EntityTeleportEvent)

Ribesg commented:
> @Ribesg Entity.teleport still seems to use a PlayerTeleportEvent.TeleportCause. This seems to tell me that we should have a unified TeleportCause api, so that Entity.teleport can have the same signature between players and entities.

I justified my choice in a previous comment. A unified TeleportCause would be a breaking change, and some of the values in this unified TeleportCause would be invalid for Players or for non-Player-Entities. There's a difference. I agree that there is still something not optimal, as most TeleportCause are shared.

We could add an Entity.teleport using the Entity TeleportCause and deprecate the old one, I think I'll do that.

> Also, currently Entity.teleport doesn't throw an event (even though we have an EntityTeleportEvent)

I'm 87% sure that this is intended. The use of the API should not throw events, there are multiple other cases where it doesn't throw events. I don't know, when a plugin plays a sound there's no NotePlayEvent. Maybe not the best example but you see what I mean :p

nightpool commented:
I don't like a Player.teleport and an Entity.teleport with confusing
signatures though. Can we make sure to at least qualify them as
Player/Entity TeleportEvent.TeleportCause in the definition? I'm afraid
that this is going to be very confusing to plugin authors trying to
implement this.

"The use of the API should not throw events" Then why do we have a
TeleportCause.PLUGIN? If you use Entity.teleport on a Player, that throws
an event? I think this code just never got updated from when
EntityTeleportEvent got added.

On Sat, May 10, 2014 at 2:24 PM, Gael Ribes notifications@github.comwrote:

> @Ribesg https://github.com/Ribesg Entity.teleport still seems to use a
> PlayerTeleportEvent.TeleportCause. This seems to tell me that we should
> have a unified TeleportCause api, so that Entity.teleport can have the same
> signature between players and entities.
> 
> I justified my choice in a previous comment. A unified TeleportCause would
> be a breaking change, and some of the values in this unified TeleportCause
> would be invalid for Players or for non-Player-Entities. There's a
> difference. I agree that there is still something not optimal, as most
> TeleportCause are shared.
> 
> We could add an Entity.teleport using the Entity TeleportCause and
> deprecate the old one, I think I'll do that.
> 
> Also, currently Entity.teleport doesn't throw an event (even though we
> have an EntityTeleportEvent)
> 
> I'm 87% sure that this is intended. The use of the API should not throw
> events, there are multiple other cases where it doesn't throw events. I
> don't know, when a plugin plays a sound there's no NotePlayEvent. Maybe not
> the best example but you see what I mean :p
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/Bukkit/Bukkit/pull/1002#issuecomment-42751682
> .

turt2live commented:
@Ribesg your testing materials appear inadequate. If possible, please update them to include the source and the compiled form of the test plugin. The test case should test all components of the PR and not just one minor part of it.

Ribesg commented:
@turt2live tell me how you would test it then, I would be happy to implement it.

turt2live commented:
The plugin does not appear to work. I also apparently glossed over the fact that a plugin can't define a custom reason, hence my original comment of "inadequate". Sorry about that. (However, maybe that should be added? See #teleport(Entity, TeleportCause) )

edit: I apologize in advance if I'm making a fool of myself. 40 something PRs later and nearly 3am do weird things to you.

Ribesg commented:
My idea was more about triggering teleports the Vanilla way, aka launching items in a portal, spawning endermen, that sort of things. The test plugin command teleporting entities was here to test the PLUGIN cause... But this doesn't trigger an event for an unknown reason https://github.com/Bukkit/CraftBukkit/blob/master/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java#L210

turt2live commented:
Considering in general plugins shouldn't be firing events due to API, is there a need for the `PLUGIN` cause if it's not used?

Ribesg commented:
It's just the same question again and again... Teleporting a Player throws an event, why not for Entities?

turt2live commented:
Then your PR needs to handle that case, as it does not right now.

Ribesg commented:
This PR was supposed to add a TeleportCause to existing EntityTeleportEvent, not to add EntityTeleportEvent calls, but if you think that it fits in this PR, I'll add it.

I think it would still be a separate issue on Jira.

turt2live commented:
Ah, fair enough. It should be fine to not fix that for the time being (misinterpreted what this PR did)

Ribesg committed:
Plugin teleporting Entities doesn't trigger an EntityTeleportEvent

Ribesg commented:
- Rebased (Something is wrong, all the commits are in the PR, will be fixed)
- Fixed TestPlugin version
- Fixed PR description (Testing Results and Materials)
- Removed EntityTeleportEvent.TeleportCause.PLUGIN
- Draft on handling out-of-portal-block teleportations ([See CB comment](https://github.com/Ribesg/CraftBukkit/commit/e4f3962799b3482065ae6b28def77763cbe4973b#commitcomment-7078101))

Ribesg commented:
- Rebase fixed.

Ribesg commented:
I now consider this PR done, waiting for other comments :-)
